### ISSUE TITLE
Invalid components.implements

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1801]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#1801|
|Type|defect|
|Created|2006-06-07 18:36:03Z|

While looking at [#11150](https://github.com/twisted/twisted/issues/11150), I found a problem in mail.mail.py:251: a call to an undefined function 'components.implements'. It looks like a deprecated call that remains here. I found 3 other occurences of this, in mail.tap.py:100 and in pair.tuntap.py:47/51.

I attach a patch removing this and replacing with 'providedBy' (as it's what I think it should do), and a test-case for mail.mail (no unittest for tap).

Attachments:

* [implements_1801.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/775/775ea0a75e300e34f175504dbe9b24e5b0da74ac/32b661a90bf257f20fdcef6874f702ce56b5966d.diff) (2619 bytes) - added by therve on 2006-06-07 18:37:24Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1801 1801
type__defect defect
reporter__therve therve
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1149705363000000 1149705363000000
changetime__1149998156000000 1149998156000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Actually it wouldn't.  Pyflakes doesn't detected errors of the form:

```python
import x
x.y()
```

where there is no such attribute ```y``` of ```x```.

However, I think an automatic pyflakes step would be useful anyway.  But how about as a pre-commit hook which rejects the commit if there is any output?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Eh wrong ticket.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

twisted.mail parts of this fixed in [branches/components.implements-1801,](branches/components.implements-1801,) please review.

pair parts of this made SEP in [#1813](https://github.com/twisted/twisted/issues/1813)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @dreid|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @exarkun|
|-|-|

This looks fine.  My only complaint about the branch is how long it took me to get through all the whitespace cleanup.  Merge away.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

(In [17168]) Merge components.implements-1801

Author: exarkun
Reviewer: dreid
Fixes [#1801](https://github.com/twisted/twisted/issues/1801)

Remove two remaining uses of t.p.c.implements, which no longer exists.  Also add
test coverage for the codepaths which were still using it and clean up a bunch of
nearby trailing whitespace.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
